File: hu/codes/python/chapter_stack_and_queue/linkedlist_stack.py
"""
File: linkedlist_stack.py
Created Time: 2022-11-29
Author: Peng Chen (pengchzn@gmail.com)
"""

import sys
from pathlib import Path

sys.path.append(str(Path(__file__).parent.parent))
from modules import ListNode


class LinkedListStack:
    """Láncolt listán alapuló verem implementáció"""

    def __init__(self):
        """Konstruktor"""
        self._peek: ListNode | None = None
        self._size: int = 0

    def size(self) -> int:
        """A verem hosszának lekérése"""
        return self._size

    def is_empty(self) -> bool:
        """Ellenőrzi, hogy a verem üres-e"""
        return self._size == 0

    def push(self, val: int):
        """Push (elem betolása)"""
        node = ListNode(val)
        node.next = self._peek
        self._peek = node
        self._size += 1

    def pop(self) -> int:
        """Pop (elem kivétele)"""
        num = self.peek()
        self._peek = self._peek.next
        self._size -= 1
        return num

    def peek(self) -> int:
        """A verem tetején lévő elem elérése"""
        if self.is_empty():
            raise IndexError("A verem üres")
        return self._peek.val

    def to_list(self) -> list[int]:
        """Lista visszaadása nyomtatáshoz"""
        arr = []
        node = self._peek
        while node:
            arr.append(node.val)
            node = node.next
        arr.reverse()
        return arr


"""Fő kód"""
if __name__ == "__main__":
    # Verem inicializálása
    stack = LinkedListStack()

    # Elemek betolása a verembe
    stack.push(1)
    stack.push(3)
    stack.push(2)
    stack.push(5)
    stack.push(4)
    print("verem =", stack.to_list())

    # A verem tetején lévő elem elérése
    peek: int = stack.peek()
    print("Verem teteje elem peek =", peek)

    # Elem kivétele a veremből
    pop: int = stack.pop()
    print("Kivett elem pop =", pop)
    print("Pop után verem =", stack.to_list())

    # A verem hosszának lekérése
    size: int = stack.size()
    print("Verem hossza size =", size)

    # Ellenőrzi, hogy üres-e
    is_empty: bool = stack.is_empty()
    print("A verem üres-e =", is_empty)

File: hu/codes/java/chapter_stack_and_queue/linkedlist_stack.java
/**
 * File: linkedlist_stack.java
 * Created Time: 2022-11-25
 * Author: krahets (krahets@163.com)
 */

package chapter_stack_and_queue;

import java.util.*;
import utils.*;

/* Láncolt lista alapú verem implementáció */
class LinkedListStack {
    private ListNode stackPeek; // Fejcsomópontot használja verem tetőként
    private int stkSize = 0; // Verem hossza

    public LinkedListStack() {
        stackPeek = null;
    }

    /* Verem hosszának lekérése */
    public int size() {
        return stkSize;
    }

    /* Verem üresség ellenőrzése */
    public boolean isEmpty() {
        return size() == 0;
    }

    /* Elem berakása a verembe */
    public void push(int num) {
        ListNode node = new ListNode(num);
        node.next = stackPeek;
        stackPeek = node;
        stkSize++;
    }

    /* Elem kivétele a veremből */
    public int pop() {
        int num = peek();
        stackPeek = stackPeek.next;
        stkSize--;
        return num;
    }

    /* Lista visszaadása nyomtatáshoz */
    public int peek() {
        if (isEmpty())
            throw new IndexOutOfBoundsException();
        return stackPeek.val;
    }

    /* Lista tömbbé konvertálása és visszaadása */
    public int[] toArray() {
        ListNode node = stackPeek;
        int[] res = new int[size()];
        for (int i = res.length - 1; i >= 0; i--) {
            res[i] = node.val;
            node = node.next;
        }
        return res;
    }
}

public class linkedlist_stack {
    public static void main(String[] args) {
        /* Verem tetején lévő elem elérése */
        LinkedListStack stack = new LinkedListStack();

        /* Elemek verembe rakása */
        stack.push(1);
        stack.push(3);
        stack.push(2);
        stack.push(5);
        stack.push(4);
        System.out.println("Stack stack = " + Arrays.toString(stack.toArray()));

        /* Lista visszaadása nyomtatáshoz */
        int peek = stack.peek();
        System.out.println("Stack top element peek = " + peek);

        /* Elem kivétele a veremből */
        int pop = stack.pop();
        System.out.println("Pop element pop = " + pop + ", after pop, stack = " + Arrays.toString(stack.toArray()));

        /* Verem hosszának lekérése */
        int size = stack.size();
        System.out.println("Stack length size = " + size);

        /* Üresség ellenőrzése */
        boolean isEmpty = stack.isEmpty();
        System.out.println("Stack is empty = " + isEmpty);
    }
}
